File: leetcode/array/723_Candy_Crush/codes/simulation.c
/**
 * Return an array of arrays of size *returnSize.
 * The sizes of the arrays are returned as *returnColumnSizes array.
 * Note: Both returned array and *columnSizes array must be malloced, assume caller calls free().
 */
#include <stdlib.h>
#include <stdbool.h>
#include <math.h>

int** candyCrush(int** board, int boardSize, int* boardColSize, int* returnSize, int** returnColumnSizes) {
    int m = boardSize, n = boardColSize[0];
    bool changed = true;

    while (changed) {
        changed = false;

        for (int i = 0; i < m; i++)
            for (int j = 0; j < n - 2; j++) {
                int v = abs(board[i][j]);
                if (v != 0 && abs(board[i][j + 1]) == v && abs(board[i][j + 2]) == v) {
                    board[i][j] = board[i][j + 1] = board[i][j + 2] = -v;
                    changed = true;
                }
            }

        for (int i = 0; i < m - 2; i++)
            for (int j = 0; j < n; j++) {
                int v = abs(board[i][j]);
                if (v != 0 && abs(board[i + 1][j]) == v && abs(board[i + 2][j]) == v) {
                    board[i][j] = board[i + 1][j] = board[i + 2][j] = -v;
                    changed = true;
                }
            }

        for (int j = 0; j < n; j++) {
            int write = m - 1;
            for (int i = m - 1; i >= 0; i--)
                if (board[i][j] > 0)
                    board[write--][j] = board[i][j];
            for (int i = write; i >= 0; i--)
                board[i][j] = 0;
        }
    }

    *returnSize = boardSize;
    *returnColumnSizes = (int*)malloc(sizeof(int) * boardSize);
    for (int i = 0; i < boardSize; i++)
        (*returnColumnSizes)[i] = boardColSize[0];
    return board;
}

File: leetcode/array/723_Candy_Crush/codes/simulation.py
from typing import List

class Solution:
    def candyCrush(self, board: List[List[int]]) -> List[List[int]]:
        m, n = len(board), len(board[0])
        changed = True

        while changed:
            changed = False
            for i in range(m):
                for j in range(n - 2):
                    val = abs(board[i][j])
                    if val != 0 and abs(board[i][j + 1]) == val and abs(board[i][j + 2]) == val:
                        board[i][j] = board[i][j + 1] = board[i][j + 2] = -val
                        changed = True
            for i in range(m - 2):
                for j in range(n):
                    val = abs(board[i][j])
                    if val != 0 and abs(board[i + 1][j]) == val and abs(board[i + 2][j]) == val:
                        board[i][j] = board[i + 1][j] = board[i + 2][j] = -val
                        changed = True

            for j in range(n):
                write = m - 1
                for i in reversed(range(m)):
                    if board[i][j] > 0:
                        board[write][j] = board[i][j]
                        write -= 1
                for i in range(write + 1):
                    board[i][j] = 0

        return board
